### PR TITLE
MAT-1873 Observation attributes

### PR DESCRIPTION
**Pull request by adongare**
MAT-1873 Observation attributes
- status
- value
- category

Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [ ] This pull request describes why these changes were made.
- [ ] This PR is into the correct branch.
- [ ] JIRA ticket for this PR:
- [ ] JIRA ticket links to this PR
- [ ] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [ ] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [ ] Tests are included and test edge cases
- [ ] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [ ] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [ ] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 
- [ ] Automated regression test(s) pass

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA test links:
- [ ] Justification for using JIRA tests:
- [ ] JIRA tests have been added to sprint

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA tests have been run and pass
- [ ] You agree with the justification for use of JIRA tests or have provided input on why you disagree

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA tests have been run and pass
- [ ] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee [hunk 1]
```diff
@@ -242,11 +248,7 @@
         title: 'clinicalStatus'
         getValue: (fhirResource) => fhirResource?.clinicalStatus?.coding?[0]
         setValue: (fhirResource, coding) =>
-          if !coding?
-            fhirResource.clinicalStatus = null
-          else
-            fhirResource.clinicalStatus = new cqm.models.CodeableConcept()
-            fhirResource.clinicalStatus.coding = [coding]
+          fhirResource.clinicalStatus = @getCodeableConceptForCoding(coding)
```
Comment (adongare) [inline app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:251]: Maybe we can refactor all setValue into single generic method and call it.
something like 
```
setResourceAttributeWithValue(fhirResource, attribute, value)
  fhirResource[attribute] = if datetime 
                    dosomeconversion
                 else if code
                    dosomeconversion
                 else if concept
                     dosomeconversion
                  else ....  
```

Comment (ghost) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:251]: I was thinking of a different approach. We could introduce an Accessor class and then we would provide a set of implementations for Accessors per each Type.

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee [hunk 2]
```diff
@@ -461,7 +431,50 @@
...
+      {
+        path: 'value'
+        title: 'value'
+        getValue: (fhirResource) =>
+          if  cqm.models.CodeableConcept.isCodeableConcept(fhirResource?.value)
+            return fhirResource.value.coding?[0]
+          else
+            fhirResource?.value
+        setValue: (fhirResource, value) =>
+          attrType = value?.constructor?.name
```
Comment (ghost) [inline app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:456]: value?.getType()

Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:456]: is that the new fix to avoid minification issue? I'll update it thanks!

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee [hunk 3]
```diff
@@ -666,11 +664,8 @@
         title: 'reasonCode',
         getValue: (fhirResource) => fhirResource?.reasonCode?[0]?.coding?[0]
         setValue: (fhirResource, coding) =>
-          if !coding?
-            fhirResource.reasonCode = null
-          else
-            fhirResource.reasonCode = [new cqm.models.CodeableConcept()]
-            fhirResource.reasonCode[0].coding = [coding]
+          codeableConcept = @getCodeableConceptForCoding(coding)
+          fhirResource.reasonCode = if codeableConcept? then [codeableConcept] else codeableConcept
```
Comment (ghost) [inline app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:668]: I'm not sure I understand that statement... does it assign a single value to an array of CodeableConcepts? Why?

File: app/assets/javascripts/views/patient_builder/inputs/string.js.coffee [hunk 4]
```diff
@@ -28,7 +28,7 @@ class Thorax.Views.InputStringView extends Thorax.Views.BonnieView
   handleInputChange: (e) ->
     inputValue = @$(e.target).val()
     if inputValue != ''
-      @value = inputValue
+      @value = cqm.models.PrimitiveString.parsePrimitive(inputValue)
```
Comment (ghost) [inline app/assets/javascripts/views/patient_builder/inputs/string.js.coffee:31]: Does it break compatibility with existing string fields?

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 3 above)
Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:668]: yeah, because `reasonCode` is array of CodeableConcept(0...* cardinality). Just refactored it. If you see the left side, it's same.

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 1 above)
Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:251]: Not sure If I understood but that will work too if that avoids lot of duplication and helps to maintain.

File: app/assets/javascripts/views/patient_builder/inputs/string.js.coffee (hunk 4 above)
Comment (adongare) [reply app/assets/javascripts/views/patient_builder/inputs/string.js.coffee:31]: I don't think. I don't see any String type usage. this is the first story to use it.

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 1 above)
Comment (ghost) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:251]: unfortunately, there are cases when the path is complex and you have to create intermediate entities

**Review by ghost — APPROVED** [review]

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 1 above)
Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:251]: oh..I see

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 2 above)
Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:456]: probably that will not work for all types. 
e.g DateTime- we might have to update this widget to return fhir datetime. Currently it is a CQL type

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee [hunk 5]
```diff
@@ -461,7 +431,50 @@
...
+        getValue: (fhirResource) =>
+          if  cqm.models.CodeableConcept.isCodeableConcept(fhirResource?.value)
+            return fhirResource.value.coding?[0]
+          else
+            fhirResource?.value
+        setValue: (fhirResource, value) =>
+          attrType = value?.constructor?.name
+          if attrType == 'DateTime'
+            fhirResource.value = @getPrimitiveDateTimeForCqlDateTime(value)
+          else if attrType == 'Coding'
```
Comment (jkotanchik-SB) [inline app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:459]: Does it also need to check for `CodeableConcept`?

**Review by jkotanchik-SB — APPROVED** [review]

File: app/assets/javascripts/helpers/data_criteria_helpers.js.coffee (hunk 5 above)
Comment (adongare) [reply app/assets/javascripts/helpers/data_criteria_helpers.js.coffee:459]: No need. view always returns Coding
